isa/kernel.org/instruction-set.rst

```diff
@@ -11,198 +27,350 @@ all of which are 64-bits wide.
 
   ======  =====  ========================================
   source  value  description
   ======  =====  ========================================
-  BPF_K   0x00   use 32-bit immediate as source operand
-  BPF_X   0x08   use 'src_reg' register as source operand
+  BPF_K   0x00   use 32-bit 'imm' value as source operand
+  BPF_X   0x08   use 'src' register value as source operand
   ======  =====  ========================================
 
-The four MSB bits store the operation code.
-
+instruction class
+  the instruction class (see `Instruction classes`_)
 
 Arithmetic instructions
 -----------------------
 
-BPF_ALU uses 32-bit wide operands while BPF_ALU64 uses 64-bit wide operands for
+Instruction class ``BPF_ALU`` uses 32-bit wide operands (zeroing the upper 32 bits
+of the destination register) while ``BPF_ALU64`` uses 64-bit wide operands for
 otherwise identical operations.
-The code field encodes the operation as below:
 
-  ========  =====  =================================================
-  code      value  description
-  ========  =====  =================================================
-  BPF_ADD   0x00   dst += src
-  BPF_SUB   0x10   dst -= src
-  BPF_MUL   0x20   dst \*= src
-  BPF_DIV   0x30   dst /= src
-  BPF_OR    0x40   dst \|= src
-  BPF_AND   0x50   dst &= src
-  BPF_LSH   0x60   dst <<= src
-  BPF_RSH   0x70   dst >>= src
-  BPF_NEG   0x80   dst = ~src
-  BPF_MOD   0x90   dst %= src
-  BPF_XOR   0xa0   dst ^= src
-  BPF_MOV   0xb0   dst = src
-  BPF_ARSH  0xc0   sign extending shift right
-  BPF_END   0xd0   byte swap operations (see separate section below)
-  ========  =====  =================================================
+Support for ``BPF_ALU`` is required in ISA version 3, and optional in earlier
+versions.
+
+   **Note**
+
+   *Clang implementation*:
+   For ISA versions prior to 3, Clang v7.0 and later can enable ``BPF_ALU`` support with
+   ``-Xclang -target-feature -Xclang +alu32``.
+
+The 4-bit 'code' field encodes the operation as follows:
+
+========  =====  =================================================
+code      value  description
+========  =====  =================================================
+BPF_ADD   0x00   dst += src
+BPF_SUB   0x10   dst -= src
+BPF_MUL   0x20   dst \*= src
+BPF_DIV   0x30   dst = (src != 0) ? (dst / src) : 0
+BPF_OR    0x40   dst \|= src
+BPF_AND   0x50   dst &= src
+BPF_LSH   0x60   dst <<= src
+BPF_RSH   0x70   dst >>= src
+BPF_NEG   0x80   dst = ~src
+BPF_MOD   0x90   dst = (src != 0) ? (dst % src) : src
+BPF_XOR   0xa0   dst ^= src
+BPF_MOV   0xb0   dst = src
+BPF_ARSH  0xc0   sign extending shift right
+BPF_END   0xd0   byte swap operations (see `Byte swap instructions`_ below)
+========  =====  =================================================
+
+where 'src' is the source operand value.
+
+Underflow and overflow are allowed during arithmetic operations,
+meaning the 64-bit or 32-bit value will wrap.  If
+eBPF program execution would result in division by zero,
```

Comment:
uBPF terminates the program instead in the case of division by zero

Reply:
That's what the text here said before LPC based on https://www.kernel.org/doc/html/latest/bpf/bpf_design_QA.html#q-why-bpf-div-instruction-doesn-t-map-to-x64-div saying:

> Q: Why BPF_DIV instruction doesn’t map to x64 div?
> A: Because if we picked one-to-one relationship to x64 it would have made it more complicated to support on arm64 and other archs. Also it needs div-by-zero **runtime check**.

The "runtime check" implies the behavior that uBPF has, similar to where https://www.kernel.org/doc/Documentation/bpf/instruction-set.rst says about legacy packet instructions:

> These instructions have an implicit program exit condition as well. When an
> eBPF program is trying to access the data beyond the packet boundary, the
> program execution will be aborted.

So I guess we should make both behaviors compliant.

Reply:
Since ubpf and rbpf both have the same graceful abort behavior, I have updated the document to include that as a valid alterative.  @qmonnet please check.

Reply:
rbpf being for me a toy project, I hardly consider it authoritative, so I'd mostly focus on uBPF. The paragraph you wrote seems good to me.

(By the way, I'm wondering if we should ping the Solana folks about the current standardisation effort. I know they forked rbpf, but I've got no clue whether they'd be interested in compliance or whether they might diverge already.)

Reply:
On the bpf list, Alexei argued that uBPF is buggy if it returns -1 and should be fixed

Reply:
> By the way, I'm wondering if we should ping the Solana folks about the current standardisation effort. I know they forked rbpf, but I've got no clue whether they'd be interested in compliance or whether they might diverge already.

Go ahead, I don't know them.

Reply:
Reverted to not mention -1 or abort, after discussion with Alan about fixing ubpf

Reply:
Regarding Solana folks: OK, I don't know them either but I'll probably try to open an issue on their rbpf fork repo to let them know.